Markdown plugins for custom admonitions

igordsm opened the pull request:
This is a mega PR that migrates many custom admonitions from React to pure python markdown extensions plus htmx/hyperscript. 

- [X] progress button
- [X] pdf and video elements
- [X] short text questions
- [ ] long text questions
- [ ] multiple choice questions
- [ ] remember question answer on reload
- [ ] send question info to report server

toshikurauchi commented on src/active_handout_plugins/counter.py:
```diff
@@ -0,0 +1,19 @@
+            if c in el.attrib['class']:
+                title = el.find("p[@class='admonition-title']")
+                self.counters[c] += 1
+                title.text = f'{title.text} {self.counters[c]}'
```
For the future: it may be good to have a way to opt-out of the counter. Or add more things to `TO_COUNT`.

toshikurauchi commented on src/active_handout_plugins/exercise.py:
```diff
@@ -0,0 +1,12 @@
+        super().__init__('exercise', [], *args, **kwargs)
+
+    def create_question_form(self, el, submission_form):
+        return '''
```
It would be nice to have this as a template file (even better if it could be overwritten by the user :)).

toshikurauchi commented on src/active_handout_plugins/pdf.py:
```diff
@@ -0,0 +1,18 @@
+            return 
+        
+        img_el = el.find("p/img")
+        if not img_el is None:
```
I prefer `if img_el is not None`, but that's fine. Just wanted to express my feelings: ![](https://media.giphy.com/media/RJAjTowsU0K1a/giphy.gif)

toshikurauchi commented on src/active_handout_plugins/question.py:
```diff
@@ -0,0 +1,108 @@
+                return
+
+        self.counter += 1
+        el.set("id", f"{cls}-{self.counter}")
```
Extract a function here (e.g. `load_id(el)`)?

toshikurauchi commented on src/active_handout_plugins/question.py:
```diff
@@ -0,0 +1,108 @@
+                el.set("id", c[3:])
+                el.attrib['class'] = el.attrib['class'].replace(c, '')
+
+        title = el.find('p/[@class="admonition-title"]')
```
Another function?

toshikurauchi commented on src/active_handout_plugins/question.py:
```diff
@@ -0,0 +1,108 @@
+            content.append(child)
+        
+        submission_form = etree.SubElement(el, 'form')
+        hs_code = '''
```
Load from file?

toshikurauchi commented on src/active_handout_theme/main.html:
```diff
@@ -4,9 +4,89 @@
+<script src="https://unpkg.com/htmx.org@1.8.0"></script>
+
+<script>
+    function keyFromElement(el) {
```
Move this to a js file?

toshikurauchi reviewed (changes requested):
Looks good. I just think some functions are a little too long (could extract some functions) and the HTML strings could be extracted to actual HTML file templates and loaded by the function.

igordsm commented:
I addressed all comment but the two I :-1:  . You are right in the sense that it's better software engineering doing that. However, I was trying to make components fairly self contained and I don't think we gain anything practical from separating into many files. IMHO we get complexity without significant benefit. @toshikurauchi thoughts?

toshikurauchi commented:
> I addressed all comment but the two I -1 . You are right in the sense that it's better software engineering doing that. However, I was trying to make components fairly self contained and I don't think we gain anything practical from separating into many files. IMHO we get complexity without significant benefit. @toshikurauchi thoughts?

You are right. It would be over engineering. I was thinking that it could be used to customize the html in specific handouts, but I don't know if it would be useful anyway. Let's keep it in the same file.
